File: lab5/haskel/prime_factors.hs
factorize :: Int -> Int -> [Int]
factorize 1 _ = []
factorize n divisor
    | n `mod` divisor == 0 = divisor : factorize (n `div` divisor) divisor
    | otherwise            = factorize n (divisor + 1)

prime_factors :: Int -> [Int]
prime_factors n
    | n <= 1    = []
    | otherwise = factorize n 2

File: lab5/prolog/prime_factors.pl
factorize(1, _, []).
factorize(N, Divisor, [Divisor|Factors]) :-
    N mod Divisor =:= 0,
    NewN is N // Divisor,
    factorize(NewN, Divisor, Factors).
factorize(N, Divisor, Factors) :-
    N mod Divisor =\= 0,
    NextDivisor is Divisor + 1,
    factorize(N, NextDivisor, Factors).

primeFactors(N, Factors) :-
    N =< 1,
    Factors = [].
primeFactors(N, Factors) :-
    N > 1,
    factorize(N, 2, Factors).